rust_src/src/time.rs

```diff
@@ -16,7 +17,62 @@ use crate::{
     remacs_sys::{lisp_time, EmacsDouble, EmacsInt},
 };
 
-const LO_TIME_BITS: i32 = 16;
+pub const LO_TIME_BITS: i32 = 16;
+
+pub type LispTime = lisp_time;
+
+impl LispTime {
+    pub fn into_vec(self, nelem: usize) -> Vec<EmacsInt> {
+        let mut v = Vec::with_capacity(nelem);
+
+        if nelem >= 2 {
+            v.push(self.hi);
+            v.push(self.lo.into());
+        }
+        if nelem >= 3 {
+            v.push(self.us.into());
+        }
+        if nelem > 3 {
+            v.push(self.ps.into());
+        }
+
+        v
+    }
+}
+
+macro_rules! return_if_different {
+    ($a:expr, $b:expr) => {{
+        let o = $a.cmp(&$b);
+        if o != Ordering::Equal {
+            return o;
+        }
+    }};
+}
+
+impl PartialEq for LispTime {
+    fn eq(&self, other: &LispTime) -> bool {
+        self.hi == other.hi && self.lo == other.lo && self.us == other.us && self.ps == other.ps
+    }
+}
+
+impl Eq for LispTime {}
+
+impl PartialOrd for LispTime {
+    fn partial_cmp(&self, other: &LispTime) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for LispTime {
+    fn cmp(&self, other: &LispTime) -> Ordering {
+        return_if_different!(self.hi, other.hi);
+        return_if_different!(self.lo, other.lo);
+        return_if_different!(self.us, other.us);
+        return_if_different!(self.ps, other.ps);
```

Comment:
This should be a [`then_with`](https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then_with) chain, i.e. `self.hi.cmp(other.hi).then_with(|| self.lo.cmp(other.lo)).then_with(…).then_with(…)`.

Reply:
Huh, TIL. Is there some reason to use `then_with` rather than `then`?

Reply:
`then` computes its argument even if the check is unneeded. By passing in closures, `then_with` will not evaluate the next comparison until it is actually needed.

Equal.then(expensive())

Equal.then_with(expensive)